(CWC-686) Agent Self Registration

lgmugnier opened the pull request:
## Description of the change

We want to move the registration process to the agent to make it more fault tolerant, robust, and simpler for the user. 

1. The agent will be able to be started using the following command:
```$ bzero-ssm-agent -apiKey "$apiKey" [optional: -org "$orgId" -orgProvider "$orgProvider" -targetName "$targetName" -envName "$environmentID" -serviceUrl "$serviceUrl"]```

2. The agent will never store the apiKey (aka `RegistrationSecret`).
3. The agent will not be required to restart once registration is complete, if it completes successfully.

Additional Small Changes
- Stop cloudwatch errors
- Changing script so that we don't restart on a clean exit code
- Changing name so that it reports as [bzero-ssm-agent] in log
- Change log so it saves as `bzero-ssm-agent.log` instead of `amazon-ssm-agent.log`
- Added the format line to the log config script so we can turn on Trace level logs

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Refactor (code change with no functionality change)
- [ ] Enhancement (minor change below the level of a feature)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Remove feature (removes a feature we don't support anymore)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related JIRA tickets

Relates to JIRA: CWC-686, CWC-687, CWC-688, CWC-689, CWC-698, CWC-679, CWC-1363

## Potential Security Impacts

Does this PR have any security impact?
Does it fix a security issue?
Does it add attack surface?
Why do we think this change is safe?

## Checklists

### Development

- [ ] Does the code changed/added as part of this pull request have test coverage?
- [ ] Do all tests related to the changed code pass in development?
- [ ] Have you ensured that at least one other person has tested your code?
- [ ] Have you tested the code?
- [ ] Have you attached any pictures (if relevant)?

Commit:
Pass in Registration Data

Registration flag now allows us to specify a bzero registration route which takes in a json argument we specify and is then checked by membership to a struct.  We store it in the vault, do the rest of the bzero stuff and move on.

Commit:
Agent Self-Registers

- Agent goes through the old process of registration with information from reg api call response
- moved orgID and orgProvider from registration flags to being set after we get the registration information response since it's included already.
- All startup messages now say bzero-ssm-agent instead of amazon-ssm-agent
- Added special error code for Bzero Registration Errors

Commit:
Added Robustness to Agent Registration

- Add retry parameter to optionally skip retries
- Add a lot of specific error messages surrounding the API call
- If a user ever tries to register and provides valid registration information, it will clear the old aws registration so that it can redo it.

Commit:
Modify service files

- Attempt restarts ONLY when there is an unclean error (like a panic or something).  This way registration can execute and exit cleanly without constant restarts.
- Only attempt to restart 3 times, we don't want our service continually running if it's unable to register or there's some flaw.

Commit:
Change AWS warn message to debug

I don't think this is ever necessary and happens every time you try to check registration before the agent has ever registered, but I didn't want to get rid of it so I set it to debug just in case it causes issues in the future.

- Changing really minor error reporting
- Fix bug where I accidentally removed a key that let the client understand the bzeroinfo output

Commit:
Maintain old registration path and add exponential backoff

- BZero inits on register even without bzero info
- Added exponential backoff to register API call

Commit:
Allow trace logs turned on

sidpremkumar reviewed (commented):
I did not get the chance to fully review, will circle back with another review

sidpremkumar reviewed (approved):
Overall looks good, need to test it once it gets to staging.

Commit:
Bump version

Commit:
Change systemd file

Change restart back to always, just in case although will not restart on a bzero reg error or reboot requested.  Restarts are still limited to 3.

Commit:
increase registration retries

Commit:
Add new registration request variable for clarity

Commit:
Revamp

Code to comply with new requirements for agent self-registration

Commit:
Add EnvId and ServiceUrl

Make sure we only assign the OrgID and OrgProvider if they were not provided by the user as opposed to overwriting any existing value.

Commit:
Add logging to registration lib

Commit:
Add omitempty to json options in registration

If user doesn't specify EnvName or EnvID, omit from registration api request

Commit:
Restart Agent after successful registration

Once an agent registers successfully, the agent will restart
